src/app/backend/handler/terminal.go

```diff
@@ -0,0 +1,264 @@
+// Called in a loop from remotecommand as long as the process is running
+func (t TerminalSession) Read(p []byte) (int, error) {
+        m, err := t.sockJSSession.Recv()
+        if err != nil {
+          return 0, err
+        }
+
+        var msg TerminalMessage
+	if err := json.Unmarshal([]byte(m), &msg); err != nil {
+          return 0, err
+	}
+
+	switch msg.Op {
+	case "stdin":
+		return copy(p, msg.Data), nil
+	case "resize":
+		t.sizeChan <- remotecommand.TerminalSize{msg.Cols, msg.Rows}
+		return 0, nil
+	default:
+		return 0, fmt.Errorf("unknown message type '%s'", msg.Op)
+	}
+}
+
+// handles process->pty stdout
+// Called from remotecommand whenever there is any output
+func (t TerminalSession) Write(p []byte) (int, error) {
+	msg, err := json.Marshal(TerminalMessage{
+          Op: "stdout",
+          Data: string(p),
+        })
+	if err != nil {
+		return 0, err
+	}
+
+        if err = t.sockJSSession.Send(string(msg)); err != nil {
+		return 0, err
+	}
+	return len(p), nil
+}
+
+// Can be used to send the user any OOB messages
+// hterm puts these in the center of the terminal
+func (t TerminalSession) Toast(p string) (error) {
+	msg, err := json.Marshal(TerminalMessage{
+          Op: "toast",
+          Data: p,
+        })
+	if err != nil {
+		return err
+	}
+
+        if err = t.sockJSSession.Send(string(msg)); err != nil {
+		return err
+	}
+	return nil
+}
+
+
+// Closes the SockJS connection and sends the status code and reason to the client
+// Can happen if the process exits or if there is an error starting up the process
+// For now the status code is unused and reason is shown to the user (unless "")
+func (t TerminalSession) Close(status uint32, reason string) {
+  t.sockJSSession.Close(status, reason)
+}
+
+// FIXME: this structure needs locking
+var terminalSessions = make(map[string]TerminalSession)
+
+// Called by net/http for any new /api/sockjs connections
+func handleTerminalSession(session sockjs.Session) {
+  var (
+    buf string
+    err error
+    msg TerminalMessage
+    terminalSession TerminalSession
+    ok bool
+  )
+
+  if buf, err = session.Recv(); err != nil {
+    log.Printf("handleTerminalSession: can't Recv: %v", err)
+    return
+  }
+
+  if err = json.Unmarshal([]byte(buf), &msg); err != nil {
+    log.Printf("handleTerminalSession: can't UnMarshal (%v): %s", err, buf)
+    return
+  }
+
+  if msg.Op != "bind" {
+    log.Printf("handleTerminalSession: expected 'bind' message, got: %s", buf)
+    return
+  }
+
+  if terminalSession, ok = terminalSessions[msg.SessionID]; !ok {
+    log.Printf("handleTerminalSession: can't find session '%s'", msg.SessionID)
+    return
+  }
+
+  terminalSession.sockJSSession = session
+  terminalSession.bound <- nil
+  terminalSessions[msg.SessionID] = terminalSession
+}
+
+// Called from main for /api/sockjs
+func CreateAttachHandler(path string) (http.Handler) {
+  options := sockjs.DefaultOptions
+  // options.Websocket = false
```

Comment:
I am not sure if if you are asking about `options` or the whole function. I've removed the variable; it was just a leftover from experiments.

If you meant the function itself: I think it's a good idea to abstract this implementation detail out of the main package (dashboard.go). It is also how the other handlers are setup:

>         http.Handle("/", handler.MakeGzipHandler(handler.CreateLocaleHandler()))
>         http.Handle("/api/", apiHandler)
>         // TODO(maciaszczykm): Move to /appConfig.json as it was discussed in #640.
>         http.Handle("/api/appConfig.json", handler.AppHandler(handler.ConfigHandler))
>         http.Handle("/api/sockjs/", handler.CreateAttachHandler("/api/sockjs"))
>         http.Handle("/metrics", prometheus.Handler())